T239598 : Allow users to customise data granularity for edits over time graph

sanyam-git opened the pull request:
with reference to [T239598](https://phabricator.wikimedia.org/T239598)
work don : 
- at first the user will be displayed the default view (best suitable).
- then the user can change the display as per his requirements.
- the AJAX request for another display is sent when user clicks the button

![dailyView](https://user-images.githubusercontent.com/43791665/73329786-e434c000-4284-11ea-9e28-d3f9cc936403.png)
![monthlyView](https://user-images.githubusercontent.com/43791665/73329787-e434c000-4284-11ea-993d-6b19e5e0b315.png)
![yearly View](https://user-images.githubusercontent.com/43791665/73329788-e4cd5680-4284-11ea-87ee-c62933a292c8.png)

sanyam-git committed:
Update graph.html

sanyam-git committed:
Update graph.js

sanyam-git committed:
Update views.py

sanyam-git committed:
Update base.py

sanyam-git commented:
with reference to [T207846 : Encode 4-byte unicode properly.](https://phabricator.wikimedia.org/T207846) 
added `'charset' : 'utf8mb4'` in database settings in settings/base.py
![Screenshot from 2020-01-30 11-43-48](https://user-images.githubusercontent.com/43791665/73332478-e00ca080-428c-11ea-8642-dc3d7b56d36c.png)

Samwalton9 commented:
Thanks! Can you split these tasks out into two different pull requests please?

sanyam-git committed:
restored base.py

added the commit to another pull request

sanyam-git commented:
@Samwalton9 , moved the task  [T207846 : Encode 4-byte unicode properly](https://phabricator.wikimedia.org/T207846) to another [pull request](https://github.com/WikipediaLibrary/hashtags/pull/57). Thanks

Samwalton9 commented on hashtagsv2/graphs/views.py:
```diff
@@ -67,9 +67,15 @@ def time_statistics_data(request):
 
     # Split by days
-    if date_range.days < 90:
+    if (date_range.days < 90 or view_type == 'daily') and view_type != 'monthly':
```
This code was a little confusing to read. I see that it works, it was just a little complicated to parse because we're doing `if view_type != 'monthly' ` here, then `if view_type != 'yearly'` next, then defaulting to years. Is `and view_type != 'monthly'` necessary? If the `view_type` isn't 'daily' then we'll go on to the next entry anyway, right? Or am I misunderstanding?

sanyam-git replied:
To use the same if-else statements for default output and user requested output, this has to be included. When the API is called for the first time on loading the page, default output will be selected on the basis of number of days.(in this case view_type is defined to prevent any error from using undefined view_type.)
```
    else:
        view_type = 'NOTA'
```

But for the subsequent times, when the API is called by user clicking buttons, it is sending the view_type in get request as a param. So this time the output should be based only on view_type not on number of days(default one). So for example if a 'monthly' view_type is requested for a hashtag with day range for 70 days, it will execute the first if statement on the basis of `date_range.days < 90`, so as to prevent this it is checking that it didn't requested 'monthly' view_type , I set an **and** condition  `view_type!='monthly`.( same reason for `view_type!='yearly'` in second elif statement.)

But as you pointed out this seems to confusing to read. The alternate can be assigning number_of_days a large value so it doesn't affect the result of else-if.

the following commit uses an alternate method https://github.com/WikipediaLibrary/hashtags/pull/55/commits/7c1229964158d25884f77b1d5385f761cb2412f8

Jain-Aditya replied:
Your alternate method looks good and readable now.

Samwalton9 reviewed (changes requested):
Looking good! I just have one question about some code that I didn't entirely understand :)

sanyam-git committed:
alternate method

Jain-Aditya commented on hashtagsv2/hashtags/static/js/graph.js:
```diff
@@ -118,47 +118,109 @@ else{
+
+    document.getElementsByName(view_type)[0].style.borderStyle = "solid";
+
+    var status = $('#' + view_type).attr("status");
```
I don't think we need the logic from line 173 to 180 because we are performing the same in the `draw_chart` function after the animation is completed.

Jain-Aditya replied:
Therefore, we can also remove the `status` attribute from the canvas element in graph.html.

sanyam-git replied:
The `status` attribute is added so the web page doesn't have to call an API every single time the user changes the view-type in same session. The `show_chart` is used to hide all chart (now download JPEG buttons also) and display only the required one by the user and `draw_chart` draws a chart from scratch, from data received through an API call.
When the page loads the default view chart get drawn and displayed. And  when user clicks other options it calls an API and draws and displays that chart. 
But the catch here is that when it draws for the first time its `status` attribute is set to _loaded_. So next time when user clicks on that option it doesn't have to call an API again, but simple hide all the other charts and displays the required one. 
I added this functionality, as I observed that for the charts with number of days over 300-400, daily- view data can take significant time to load (observation form live project). And loading all data (three view-types) together also doesn't seems to be a smart option.

Regarding the logic from line 173-180. Earlier there was only one Download JPEG button and it needs to be changed regularly with the change in view-type by user. Therefore line 173-180 are used again. Now in the last 2 commits, I added three different buttons for different views, they will be displayed and hidden same as the respective charts.These commits will make the code more readable in my understanding.

Jain-Aditya replied:
Ohh yes, keeping `status` attribute would be fine. I just meant to remove the repeated logic behind the JPEG download, but now you have removed it in the latest commit. So it is fine. :)

sanyam-git replied:
Yep, earlier there was only one button, so the logic needs to be repeated every time, when the view is changed . But now I have added three buttons, so need of repetition. :)

Jain-Aditya replied:
Great work! maybe I misunderstood it before. But now it looks good.

sanyam-git committed:
Update graph.html

sanyam-git committed:
Update graph.js

Jain-Aditya commented:
At this point, it looks like this and the daily and yearly charts doesn't show up.
![image](https://user-images.githubusercontent.com/34097318/73610661-dc1bae00-45ff-11ea-89ca-e5eda2e4902d.png)
Your changes were fine 2 commits behind where I left my previous comment. You just had to remove the line numbers I mentioned.

sanyam-git commented:
> At this point, it looks like this and the daily and yearly charts doesn't show up.
> ![image](https://user-images.githubusercontent.com/34097318/73610661-dc1bae00-45ff-11ea-89ca-e5eda2e4902d.png)
> Your changes were fine 2 commits behind where I left my previous comment. You just had to remove the line numbers I mentioned.

I was getting the same display, just hard refresh the page to load the latest javascript changes and the display will be fine.

Jain-Aditya commented:
> > At this point, it looks like this and the daily and yearly charts doesn't show up.
> > ![image](https://user-images.githubusercontent.com/34097318/73610661-dc1bae00-45ff-11ea-89ca-e5eda2e4902d.png)
> > Your changes were fine 2 commits behind where I left my previous comment. You just had to remove the line numbers I mentioned.
> 
> I was getting the same display, just hard refresh the page and to load the latest javascript changes.

Yup, working fine!

Jain-Aditya commented:
Overall it is looking good! Also could you please modify the CSV download view as per the graphs? Currently it gives edits per day irrespective of the view type.

sanyam-git committed:
updated tests for upcoming commits

sanyam-git committed:
Update graph.html

sanyam-git committed:
Update graph.js

sanyam-git committed:
Update views.py

sanyam-git commented:
@Jain-Aditya I have added the CSV functionality and also added tests for the same.
The checks for my first and second commit looks failed maybe because I committed the remaining changes after them. Next time I will club all  the changes in one commit.
But as I see all the changes are checked/passed in the last commit.
I think it is good to go. :)

Jain-Aditya commented:
Thanks! I see there is lot of code repeated in the `time_csv` view and `time_statistics_data` view. What you could do here is internally call an API in the csv view by passing view_type parameter(You can use `requests` library) and then create a CSV out of it. This will reduce a lot of lines of code. Could you try that?

sanyam-git commented:
Great idea ! I will try to implement it. Thanks.

sanyam-git committed:
Update graph.html

sanyam-git committed:
Update graph.js

sanyam-git committed:
Update views.py

sanyam-git commented:
Implemented the code refactor.

Jain-Aditya commented on hashtagsv2/graphs/views.py:
```diff
@@ -216,30 +230,22 @@ def projects_csv(request):
     writer = csv.writer(response)
     writer.writerow([
         # Translators: Date on which edit is made.
         _('Date'),
```
Just a small change and it will be good to go. CSV header should also be changed to `month`, `date` and `year` accordingly.

sanyam-git replied:
done :+1:

sanyam-git committed:
Update views.py

Jain-Aditya commented:
Looking good! thanks @sanyam-git  for your valuable work on this task.
